### ISSUE TITLE
No 'thread support' reported on static openssl-1.1.1j

### ISSUE THREAD

**Issue by pkreuzt** [openssl/openssl#14574]
Crossbuilding openssl-1.1.1j on linux for ming32/ming64 target, statically. Used same options that worked on previous 1.1.1i version:

`$ LDFLAGS="-static -static-libgcc" ./Configure --cross-compile-prefix=x86_64-w64-mingw32- no-shared no-zlib no-asm --prefix="$HOME/openssl/install" -static mingw64`

But on 1.1.1j, despite no error at build, trying to link another program to openssl reports 'no thread support'. On installed `opensslconf.h` there should be these lines:

```
#ifndef OPENSSL_THREADS
# define OPENSSL_THREADS
#endif
```

that were present on 1.1.1i build, but missing on 1.1.1j. Skipping checks on affected programs make them build and run, but haven't tested them too much.

**Comment by pkreuzt** [comment]
Problem still present on openssl-1.1.1k .

**Comment by Jigsaw52** [comment]
I've looked into this and the issue was introduced with the following commit: https://github.com/openssl/openssl/commit/cfd7225fbb9507b2e443a494459bdaab5236d29d

By reverting this commit and building openssl 1.1.1j, I now have OPENSSL_THREADS defined in opensslconf.h.

**Comment by t8m** [comment]
The problem is that at least on some platforms linking libc statically implies no threading support. Possibly this is not a problem on mingw.

**Comment by katvillariba** [comment]
In our situation, we enabled the --static and --threads option, but the --threads option was being ignored. And instead, our static library would be configured with nothreads support. As a result, our multi-threaded application would have random crashes inside OpenSSL, and we only figured out the reason why with the help from #18748 . To avoid the crash, we followed the workaround by @Jigsaw52 and that eliminated the crashes. So think a fix is needed to honor the --threads option even when --static is enabled.

**Comment by t8m** [comment]
I do not think this will be resolved on 1.1.1 branch - it is security fixes only.

It is also rather a feature request to be able to support threads with -static.

**Comment by t8m** [comment]
@levitte what do you think about this?

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by tristan957** [comment]
@nhorman what would it take for this thread to move into the "active" state? I can't really figure out if this is intended or not from the discussion in the thread so far.

I happen to be someone interested in a static build + threading support on glibc. If the answer to make this active again is providing a PR to fix the Configure script, I am happy to do so. I'm not aware of any other issues that might be lying under the covers if I do make that happen however.

Given this response from t8m:

> The problem is that at least on some platforms linking libc statically implies no threading support. Possibly this is not a problem on mingw.

Perhaps it could be an opt-in? If no-shared is set, disable threading, unless threading is explicitly requested.

**Comment by nhorman** [comment]
@tristan957 , you're adding a comment here is absolutely sufficient.  The warnings about closure were only made because the bug had no commentary for a year, and we're trying to clean things up.  If you are willing and interested in working on this, I'll move it to a community issues and assign you right now.

Thank you!

**Comment by tristan957** [comment]
Thanks. Let me talk to my company to see if they can assign me some time to work on this.

**Comment by t8m** [comment]
I am resolving this as won't fix. Linking libc statically with thread support enabled is problematic.

**Comment by kathyrene-villariba** [comment]
Why is linking libc statically with thread support problematic?

**Comment by icedream** [comment]
Tor's fully static build is no longer possible without patching as a result of the commit as it relies on OpenSSL providing threading functionality even for that. Stumbled upon this while trying to find a fix.
